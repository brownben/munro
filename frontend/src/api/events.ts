import { getData, postData, putData, deleteData } from './requests'

export const getEvent = (id: string): Promise<LeagueEvent | null> =>
  getData<LeagueEvent>({
    apiLocation: `/api/events/${id}`,
    customErrorMessage: 'Problem Fetching Event Details',
  })

export const getEvents = (): Promise<LeagueEvent[] | null> =>
  getData<LeagueEvent[]>({
    apiLocation: '/api/events',
    customErrorMessage: 'Problem Fetching Events',
  })

export const getLeagueEvents = (
  league: string,
  uploadKey?: boolean
): Promise<LeagueEvent[] | null> => {
  const apiLocation = uploadKey
    ? `/api/leagues/${league}/events/uploadKey`
    : `/api/leagues/${league}/events`

  return getData<LeagueEvent[]>({
    apiLocation,
    customErrorMessage: 'Problem Fetching Events',
  })
}

export const getLatestResults = (): Promise<LeagueEvent[] | null> =>
  getData<LeagueEvent[]>({
    apiLocation: '/api/events/latest-results',
    customErrorMessage: 'Problem Fetching Events',
  })

export const createEvent = (data: LeagueEvent): Promise<ServerMessage | null> =>
  postData<ServerMessage>({
    apiLocation: `/api/events`,
    data,
    customErrorMessage: 'Problem Creating Event',
    customSuccessMessage: `Event \`${data.name}\` Created`,
    customErrorHandler: true,
  })

export const updateEvent = (data: LeagueEvent): Promise<ServerMessage | null> =>
  putData<ServerMessage>({
    apiLocation: `/api/events/${data.id}`,
    data,
    customErrorMessage: 'Problem Updating Event',
    customSuccessMessage: `Event \`${data.name}\` Updated`,
    customErrorHandler: true,
  })

export const deleteEvent = (
  id: string,
  name: string
): Promise<ServerMessage | null> =>
  deleteData<ServerMessage>({
    apiLocation: `/api/events/${id}`,
    customErrorMessage: 'Problem Deleting Event',
    customSuccessMessage: `Event \`${name}\` Deleted`,
  })
