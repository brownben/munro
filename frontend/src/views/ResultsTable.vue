<!--
  League Results for Course

  Display results for course in dynamic table with searching and sorting. Click on the relevent cell in the header
  to sort/ change sort direction/ preference
  On mobile show overview with name, class, total points and when clicked reveal points for each event.
  Points not included in total are strikethrough style
-->

<template>
  <div>
    <vue-headful
      :title="'Munro - ' + $route.params.name + ' - ' + $route.params.course + ' Results'"
      :description="'Results from the ' + $route.params.course + ' course of the ' + $route.params.name + ' league on Munro, the Fast and Easy Results System for Orienteering Leagues. A simple way to calculate the results for orienteering leagues, with search and sort features'"
      :url="'https://munro-leagues.herokuapp.com/leagues'+$route.params.name+'/results/'+$route.params.course"
      :head="{
        'meta': {name: 'robots', content:'all'},
      }"
    />
    <h1 id="league-title">
      <router-link :to="'/leagues/'+$route.params.name">{{ $route.params.name }}</router-link>
      - {{ $route.params.course }}
    </h1>
    <filter-menu @changed="filterChanged" />
    <transition name="shrink">
      <div v-show="filteredResults && filteredResults.length > 0">
        <table>
          <thead>
            <tr>
              <th @click="sortBy('position')">
                <p>Pos.</p>
                <up-down-arrow :ascending="ascendingSort" :active="sortedBy === 'position'" />
              </th>
              <th @click="sortBy('name')">
                <p>Name</p>
                <up-down-arrow :ascending="ascendingSort" :active="sortedBy === 'name'" />
              </th>
              <th @click="sortBy('age')">
                <p>Class</p>
                <up-down-arrow :ascending="ascendingSort" :active="sortedBy === 'age'" />
              </th>
              <th class="club" @click="sortBy('club')">
                <p>Club</p>
                <up-down-arrow :ascending="ascendingSort" :active="sortedBy === 'club'" />
              </th>
              <th @click="sortBy('totalPoints')">
                <p>Points</p>
                <up-down-arrow :ascending="ascendingSort" :active="sortedBy === 'totalPoints'" />
              </th>
              <template v-if="!smallWindow">
                <th
                  v-for="event of eventsWithResults"
                  :key="eventsWithResults.indexOf(event)"
                  @click="sortBy('points-' + eventsWithResults.indexOf(event))"
                >
                  <p>{{ eventsWithResults.indexOf(event) + 1 }}</p>
                  <span>{{ event.name }}</span>
                  <up-down-arrow
                    :ascending="ascendingSort"
                    :active="sortedBy === ('points-' + eventsWithResults.indexOf(event))"
                  />
                </th>
              </template>
              <td v-else />
            </tr>
          </thead>
          <tbody is="transition-group" name="fade">
            <template v-for="result of filteredResults">
              <tr
                :key="result.name"
                :class="{ striped: filteredResults.indexOf(result) % 2 === 0 }"
                class="normal-table-row"
                @click="toggleRow(filteredResults.indexOf(result))"
              >
                <!-- :class - If odd number results, add background color to give stripe effect to make it easier to read -->
                <td>{{ result.position }}</td>
                <td>{{ result.name }}</td>
                <td>{{ result.ageClass }}</td>
                <td class="club">{{ result.club }}</td>
                <td>{{ result.totalPoints }}</td>
                <template v-if="!smallWindow">
                  <td
                    v-for="event of eventsWithResults"
                    :key="eventsWithResults.indexOf(event)"
                    :class="{ strikethrough: !result.largestPoints.includes(eventsWithResults.indexOf(event)) }"
                  >
                    {{ result.points[eventsWithResults.indexOf(event)] }}
                  </td>
                </template>
                <td v-else>
                  <svg
                    v-if="!openedRows.includes(filteredResults.indexOf(result))"
                    width="16"
                    height="16"
                    viewBox="0 0 24 24"
                  >
                    <path d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z" />
                    <path d="M0 0h24v24H0z" fill="none" />
                  </svg>
                  <svg v-else width="16" height="16" viewBox="0 0 24 24">
                    <path d="M12 8l-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z" />
                    <path d="M0 0h24v24H0z" fill="none" />
                  </svg>
                </td>
              </tr>
              <tr
                v-if="smallWindow && openedRows.includes(filteredResults.indexOf(result))"
                :key="result.name + '-mobile'"
                :class="{ striped: filteredResults.indexOf(result) % 2 === 0 }"
                class="mobile-table-expansion"
              >
                <!-- :class - If odd number results, add background color to give stripe effect to make it easier to read -->
                <td colspan="100%">
                  <p v-for="event of eventsWithResults" :key="eventsWithResults.indexOf(event)">
                    {{ event.name }}:
                    <span
                      :class="{strikethrough: !result.largestPoints.includes(eventsWithResults.indexOf(event))}"
                    >{{ result.points[eventsWithResults.indexOf(event)] }}</span>
                  </p>
                </td>
              </tr>
            </template>
          </tbody>
        </table>
        <div v-if="otherCourses" class="card">
          <h2>Results for Other Courses</h2>
          <div>
            <router-link
              v-for="course in otherCourses"
              :key="course"
              :to="'/leagues/'+$route.params.name+'/results/'+course"
              class="button"
            >
              {{ course }}
            </router-link>
          </div>
        </div>
      </div>
    </transition>
    <!-- If no results show, message -->
    <h2 v-if="!found">Sorry, No Results Could Be Found</h2>
  </div>
</template>

<script>
import axios from 'axios'
import FilterMenu from '@/components/FilterMenu'
import UpDownArrow from '@/components/UpDownArrows'

export default {
  components: {
    FilterMenu: FilterMenu,
    UpDownArrow: UpDownArrow,
  },

  data: () => ({
    smallWindow: false,
    found: true,
    openedRows: [],
    rawResults: [],
    events: [],
    otherCourses: [],
    sortedBy: 'position',
    ascendingSort: false,
    filterPreferences: {
      name: '',
      club: '',
      minAge: 0,
      maxAge: 100,
      male: true,
      female: true,
    },
  }),

  computed: {
    resultsWithAgeClassSplit: function () {
      // Split age class into age and gender to allow easy sorting
      return this.rawResults.map(result => {
        if (result.ageClass) {
          result.gender = result.ageClass[0]
          result.age = parseInt(result.ageClass.slice(1))
        }
        return result
      })
    },

    sortedResults: function () {
      // Sort results by preference
      let property = this.sortedBy
      if (this.sortedBy.includes('points-')) property = parseInt(property.split('-')[1])
      return this.sort(
        this.resultsWithAgeClassSplit,
        property,
        this.ascendingSort,
        this.sortedBy.includes('points-'))
    },

    filteredResults: function () {
      // Filter results by preferences
      return this.sortedResults
        .filter(result => result.name.match(new RegExp(this.filterPreferences.name, 'i'))) // Filter by Name Case Insensitive
        .filter(result => result.club.match(new RegExp(this.filterPreferences.club, 'i'))) // Filter by Club Case Insensitive
        .filter(result => {
          if (this.filterPreferences.maxAge === 100 && this.filterPreferences.minAge === 0) return true
          else {
            return this.filterPreferences.minAge <= result.age && result.age <= this.filterPreferences.maxAge
          }
        }) // If age in range
        .filter(result => (this.filterPreferences.male && this.filterPreferences.female) ||
          (this.filterPreferences.male && result.gender === 'M') ||
          (this.filterPreferences.female && result.gender === 'W')) // Filter by Gender
    },

    eventsWithResults: function () {
      // Get events with results
      return this.events.filter(event => event.resultUploaded)
    },
  },

  // If route changes without reload (if only course parameter changes)
  watch: {
    $route: async function () {
      this.rawResults = []
      await this.getResults()
      await this.getEventList()
      this.getOtherCourses()
    },
  },

  // On load
  mounted: async function () {
    // Mobile resize watcher
    window.addEventListener('resize', this.handleResize)
    this.handleResize()

    // Fetch Data
    await this.getResults()
    await this.getEventList()
    this.getOtherCourses()
  },

  destroyed () {
    window.removeEventListener('resize', this.handleResize)
  },

  methods: {
    handleResize: function () {
      if (window.innerWidth > 900) this.smallWindow = false
      else this.smallWindow = true
    },

    getResults: function () {
      return axios
        .get('/api/leagues/' + this.$route.params.name + '/results/' + this.$route.params.course)
        .then(response => {
          if (response.data.length > 0) this.rawResults = response.data
          else this.found = false
        })
        .catch(() => { this.found = false })
    },

    getOtherCourses: function () {
      return axios
        .get('/api/leagues/' + this.$route.params.name)
        .then(response => {
          if (response.data.courses) this.otherCourses = response.data.courses.filter(course => course !== this.$route.params.course)
          else this.otherCourses = false
        })
        .catch(() => { this.otherCourses = false })
    },

    getEventList: function () {
      return axios
        .get('/api/leagues/' + this.$route.params.name + '/events')
        .then(response => { this.events = response.data })
        .catch(() => this.$messages.addMessage('Problem Fetching List of Events'))
    },

    sort: function (array, property, ascending = true, byPoints = false) {
      // Selection Sort using Single List for Sorting Results
      let sortFunction
      if (byPoints) {
        sortFunction = (a, b) => {
          if (a.points[property] === b.points[property]) return 0
          else if (a.points[property] === null) return 1
          else if (b.points[property] === null) return -1
          else if (a.points[property] < b.points[property]) return 1
          else return -1
        }
      }
      else {
        sortFunction = (a, b) => {
          if (a[property] === b[property]) return 0
          else if (a[property] === null || a[property] === undefined) return 1
          else if (b[property] === null || b[property] === undefined) return -1
          else if (a[property] < b[property]) return 1
          else return -1
        }
      }
      if (ascending) return array.sort(sortFunction)
      else return array.sort(sortFunction).reverse()
    },

    sortBy: function (sortBy) {
      // Change what property it is sorted by
      this.openedRows = []
      // If it is a different property, make it sort ascending else change direction of sort
      if (sortBy !== this.sortedBy) this.ascendingSort = false
      else this.ascendingSort = !this.ascendingSort
      this.sortedBy = sortBy
    },

    filterChanged: function (data) {
      // Update data of view if Filter Menu emits a change
      this.filterPreferences.name = data.name
      this.filterPreferences.club = data.club
      if (data.minAge === '') this.filterPreferences.minAge = 0
      else this.filterPreferences.minAge = data.minAge
      if (data.maxAge === '') this.filterPreferences.maxAge = 100
      else this.filterPreferences.maxAge = data.maxAge
      this.filterPreferences.male = data.male
      this.filterPreferences.female = data.female
    },

    toggleRow: function (id) {
      // Shows the detailed points view when the row for the results is clicked on a mobile
      const index = this.openedRows.indexOf(id)
      if (index > -1) this.openedRows.splice(index, 1)
      else this.openedRows.push(id)
    },
  },
}
</script>

<style lang="stylus" scoped>
@import '../assets/styles/helpers.styl'
@import '../assets/styles/inputs.styl'
@import '../assets/styles/table.styl'

#router-view
  padding: 1rem 7.5%

  @media (max-width: 900px)
    padding: 1rem 5%

#league-title
  padding: 0 0 0.75rem

  a
    color: main-color
    text-decoration: underline

table tr th p
  display: inline-flex
  font-weight: 400

h1
  a
    text-decoration: none

th
  position: relative

  span
    position: absolute
    bottom: 80%
    left: 50%
    margin-left: -45px
    padding: 0.25rem 0.5rem
    width: 80px
    background-color: white
    color: purple-500
    opacity: 0
    transition: 0.3s
    box-shadow(1)

th:hover > span
  opacity: 1

@media (max-width: 500px)
  .club
    display: none

.card
  box-shadow(1)
  margin-top: 1.5rem
  padding: 1rem

  a
    margin-top: 0.5rem
    margin-left: 0.4rem

    &:first-child
      margin-left: 0

.strikethrough
  text-decoration: line-through

.fade-enter-active, .fade-leave-active
  transition: opacity 0.5s

.fade-enter, .fade-leave-to
  opacity: 0
</style>
