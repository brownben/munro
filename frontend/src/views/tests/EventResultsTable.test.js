import axios from 'axios'
import VueRouter from 'vue-router'
import { createLocalVue, mount, shallowMount } from '@vue/test-utils'

import ResultsTable from '@/views/EventResultsTable.vue'
import { sampleThreeEvents } from '@/tests/test data/events'
import {
  sampleSingleResult,
  sampleThreeResults,
  sampleThreeResultsSortedByAgeClass,
  sampleThreeResultsWithSplitAgeClass,
} from '@/tests/test data/results'

// A test is modifying sampleThreeResults so create backup copy
const sampleThreeResultsCopy = JSON.parse(JSON.stringify(sampleThreeResults))

jest.mock('axios')
const flushPromises = () =>
  new Promise((resolve) => global.setImmediate(resolve))
beforeEach(() => {
  jest.resetAllMocks()
  axios.get.mockResolvedValue()
})

test('Is a Vue Instance', () => {
  const wrapper = shallowMount(ResultsTable, {
    mocks: {
      $route: { params: { name: '', course: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly + Fetches Data', async () => {
  const mockGetResults = jest.fn().mockResolvedValue()
  const wrapper = shallowMount(ResultsTable, {
    mocks: {
      $route: { params: { name: '', course: '' } },
      $messages: { addMessage: jest.fn() },
    },
    methods: {
      getResults: mockGetResults,
    },
    computed: {
      eventsWithResults: () => sampleThreeEvents,
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  expect(mockGetResults).toHaveBeenCalledTimes(1)
  expect(mockGetResults).toHaveBeenLastCalledWith()
  wrapper.setData({
    rawResults: sampleThreeResults,
    event: sampleThreeEvents[0],
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Elapsed Time', () => {
  const wrapper = mount(ResultsTable, {
    mocks: {
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.vm.elapsedTime(0)).toBe('')
  expect(wrapper.vm.elapsedTime(1)).toBe('00:01')
  expect(wrapper.vm.elapsedTime(45)).toBe('00:45')
  expect(wrapper.vm.elapsedTime(60)).toBe('01:00')
  expect(wrapper.vm.elapsedTime(83)).toBe('01:23')
  expect(wrapper.vm.elapsedTime(1023)).toBe('17:03')
  expect(wrapper.vm.elapsedTime(3600)).toBe('60:00')
  expect(wrapper.vm.elapsedTime(5117)).toBe('85:17')
})

test('Results With Age Class Split', () => {
  const wrapper = shallowMount(ResultsTable, {
    mocks: {
      $route: { params: { name: '', course: '' } },
      $messages: { addMessage: jest.fn() },
    },
    computed: {
      filteredResults: jest.fn(() => []),
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ rawResults: [] })
  expect(wrapper.vm.resultsWithAgeClassSplit).toEqual([])
  wrapper.setData({ rawResults: [{ name: 'M45' }] })
  expect(wrapper.vm.resultsWithAgeClassSplit).toEqual([
    { name: 'M45', gender: '' },
  ])
  wrapper.setData({ rawResults: [{ ageClass: 'M45' }] })
  expect(wrapper.vm.resultsWithAgeClassSplit).toEqual([
    { ageClass: 'M45', age: 45, gender: 'M' },
  ])
  wrapper.setData({ rawResults: [{ ageClass: 'W100' }] })
  expect(wrapper.vm.resultsWithAgeClassSplit).toEqual([
    { ageClass: 'W100', age: 100, gender: 'W' },
  ])
  wrapper.setData({ rawResults: [{ ageClass: 'M5' }] })
  expect(wrapper.vm.resultsWithAgeClassSplit).toEqual([
    { ageClass: 'M5', age: 5, gender: 'M' },
  ])
  wrapper.setData({
    rawResults: [
      { ageClass: 'M5' },
      { ageClass: 'W100' },
      { ageClass: 'M45' },
      { ageClass: 'W16' },
    ],
  })
  expect(wrapper.vm.resultsWithAgeClassSplit).toEqual([
    { ageClass: 'M5', age: 5, gender: 'M' },
    { ageClass: 'W100', age: 100, gender: 'W' },
    { ageClass: 'M45', age: 45, gender: 'M' },
    { ageClass: 'W16', age: 16, gender: 'W' },
  ])
})

test('Refresh Data on Route Change', async () => {
  const mockGetResults = jest.fn().mockResolvedValue()
  const mockGetEvent = jest.fn().mockResolvedValue()
  const localVue = createLocalVue()
  localVue.use(VueRouter)
  const router = new VueRouter({ routes: [{ path: '/' }, { path: '/other' }] })
  shallowMount(ResultsTable, {
    localVue,
    router,
    mocks: {
      $messages: { addMessage: jest.fn() },
    },
    methods: {
      getResults: mockGetResults,
      getEvent: mockGetEvent,
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  expect(mockGetResults).toHaveBeenCalledTimes(0)
  expect(mockGetEvent).toHaveBeenCalledTimes(0)
  router.push('/other')
  await flushPromises()
  expect(mockGetResults).toHaveBeenCalledTimes(1)
  expect(mockGetResults).toHaveBeenLastCalledWith()
  expect(mockGetEvent).toHaveBeenCalledTimes(1)
  expect(mockGetEvent).toHaveBeenLastCalledWith()
})

test('Filter Changed', () => {
  const wrapper = shallowMount(ResultsTable, {
    mocks: {
      $route: { params: { name: '', course: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.vm.filterChanged({
    name: 'a',
    club: 'b',
    minAge: 10,
    maxAge: 90,
    male: true,
    female: false,
  })
  expect(wrapper.vm.filterPreferences.name).toBe('a')
  expect(wrapper.vm.filterPreferences.club).toBe('b')
  expect(wrapper.vm.filterPreferences.minAge).toBe(10)
  expect(wrapper.vm.filterPreferences.maxAge).toBe(90)
  expect(wrapper.vm.filterPreferences.male).toBeTruthy()
  expect(wrapper.vm.filterPreferences.female).toBeFalsy()
  wrapper.vm.filterChanged({
    name: 'a',
    club: 'b',
    minAge: '',
    maxAge: '',
    male: true,
    female: false,
  })
  expect(wrapper.vm.filterPreferences.minAge).toBe(0)
  expect(wrapper.vm.filterPreferences.maxAge).toBe(100)
})

test('Sort By - Same Property', () => {
  const wrapper = shallowMount(ResultsTable, {
    mocks: {
      $route: { params: { name: '', course: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ sortedBy: '2', ascendingSort: false })
  wrapper.vm.sortBy('2')
  expect(wrapper.vm.ascendingSort).toBeTruthy()
  expect(wrapper.vm.sortedBy).toEqual('2')
})

test('Sort By - Different Property', () => {
  const wrapper = shallowMount(ResultsTable, {
    mocks: {
      $route: { params: { name: '', course: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ sortedBy: '1', ascendingSort: false })
  wrapper.vm.sortBy('2')
  expect(wrapper.vm.ascendingSort).toBeFalsy()
  expect(wrapper.vm.sortedBy).toEqual('2')
})

test('Sort - Ascending', () => {
  const wrapper = shallowMount(ResultsTable, {
    mocks: {
      $route: { params: { name: '', course: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.vm.sort([], 'name')).toEqual([])
  expect(wrapper.vm.sort([], 'club')).toEqual([])
  expect(wrapper.vm.sort(sampleSingleResult, 'name')).toEqual(
    sampleSingleResult
  )
  expect(wrapper.vm.sort(sampleSingleResult, 'position')).toEqual(
    sampleSingleResult
  )
  expect(wrapper.vm.sort(sampleThreeResults, 'position')).toEqual(
    sampleThreeResults
  )
  expect(wrapper.vm.sort(sampleThreeResultsCopy, 'ageClass')).toEqual(
    sampleThreeResultsSortedByAgeClass.reverse()
  )
})

test('Sort - Descending', () => {
  const wrapper = shallowMount(ResultsTable, {
    mocks: {
      $route: { params: { name: '', course: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.vm.sort([], 'name', false)).toEqual([])
  expect(wrapper.vm.sort([], 'club', false)).toEqual([])
  expect(wrapper.vm.sort(sampleSingleResult, 'name', false)).toEqual(
    sampleSingleResult
  )
  expect(wrapper.vm.sort(sampleSingleResult, 'position', false)).toEqual(
    sampleSingleResult
  )
  expect(wrapper.vm.sort(sampleThreeResults, 'position', false)).toEqual(
    sampleThreeResults
  )
  expect(wrapper.vm.sort(sampleThreeResultsCopy, 'ageClass', false)).toEqual(
    sampleThreeResultsSortedByAgeClass.reverse()
  )
})

test('Filtered Results', () => {
  const wrapper = shallowMount(ResultsTable, {
    mocks: {
      $route: { params: { name: '', course: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ rawResults: sampleThreeResults, sortedBy: 'position' })
  wrapper.setData({ filterPreferences: { male: true, female: false } })
  expect(wrapper.vm.filteredResults.length).toBe(2)
  wrapper.setData({ filterPreferences: { male: false, female: true } })
  expect(wrapper.vm.filteredResults.length).toBe(1)
  wrapper.setData({ filterPreferences: { male: true, female: true } })
  expect(wrapper.vm.filteredResults.length).toBe(3)
  wrapper.setData({ filterPreferences: { male: false, female: false } })
  expect(wrapper.vm.filteredResults.length).toBe(0)
  wrapper.setData({ filterPreferences: { male: true, female: true } })
  wrapper.setData({ filterPreferences: { minAge: 22, maxAge: 60 } })
  expect(wrapper.vm.filteredResults.length).toBe(1)
  expect(wrapper.vm.filteredResults).toEqual([sampleThreeResults[0]])
  wrapper.setData({ filterPreferences: { minAge: 21, maxAge: 60 } })
  expect(wrapper.vm.filteredResults.length).toBe(2)
  wrapper.setData({ filterPreferences: { minAge: 21, maxAge: 65 } })
  expect(wrapper.vm.filteredResults.length).toBe(3)
  wrapper.setData({ filterPreferences: { minAge: 0, maxAge: 100 } })
  wrapper.setData({ filterPreferences: { name: 'Blu' } })
  expect(wrapper.vm.filteredResults.length).toBe(2)
  wrapper.setData({ filterPreferences: { name: 'bLU' } })
  expect(wrapper.vm.filteredResults.length).toBe(2)
  wrapper.setData({ filterPreferences: { name: 'BLU' } })
  expect(wrapper.vm.filteredResults.length).toBe(2)
  wrapper.setData({ filterPreferences: { name: '', club: 'ESOC' } })
  expect(wrapper.vm.filteredResults.length).toBe(2)
  wrapper.setData({ filterPreferences: { club: 'esoc' } })
  expect(wrapper.vm.filteredResults.length).toBe(2)
  wrapper.setData({ filterPreferences: { club: 'In' } })
  expect(wrapper.vm.filteredResults.length).toBe(1)
})

test('Sorted Results', () => {
  const wrapper = shallowMount(ResultsTable, {
    mocks: {
      $route: { params: { name: '', course: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ rawResults: [], sortedBy: 'position', ascendingSort: true })
  expect(wrapper.vm.sortedResults).toEqual([])
  wrapper.setData({ rawResults: [], sortedBy: 'points-1', ascendingSort: true })
  expect(wrapper.vm.sortedResults).toEqual([])
  wrapper.setData({
    rawResults: [],
    sortedBy: 'points-2',
    ascendingSort: false,
  })
  expect(wrapper.vm.sortedResults).toEqual([])
  wrapper.setData({ rawResults: [] })
  wrapper.setData({
    rawResults: sampleThreeResultsCopy,
    sortedBy: 'position',
    ascendingSort: true,
  })
  expect(wrapper.vm.sortedResults).toEqual(
    sampleThreeResultsWithSplitAgeClass.slice().reverse()
  )
  wrapper.setData({ rawResults: [] })
  wrapper.setData({
    rawResults: sampleThreeResultsCopy,
    sortedBy: 'position',
    ascendingSort: false,
  })
  expect(wrapper.vm.sortedResults).toEqual(sampleThreeResultsWithSplitAgeClass)
})

test('Get Results - Correct API Call', async () => {
  const wrapper = shallowMount(ResultsTable, {
    mocks: {
      $route: { params: { event: 'EVENT' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue()
  await wrapper.vm.getResults()
  expect(axios.get).toHaveBeenCalledTimes(1)
  expect(axios.get).toHaveBeenLastCalledWith('/api/events/EVENT/results')
})

test('Get Results - Success - Results', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(ResultsTable, {
    mocks: {
      $route: { params: { event: 'EVENT' } },
      $messages: { addMessage: mockAddMessage },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue({ data: sampleThreeResults })
  await wrapper.vm.getResults()
  expect(mockAddMessage).toHaveBeenCalledTimes(0)
  expect(wrapper.vm.rawResults).toEqual(sampleThreeResults)
})

test('Get Results - Success - No Results', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(ResultsTable, {
    mocks: {
      $route: { params: { event: 'EVENT' } },
      $messages: { addMessage: mockAddMessage },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue({ data: [] })
  await wrapper.vm.getResults()
  expect(wrapper.vm.rawResults).toEqual([])
  expect(mockAddMessage).toHaveBeenCalledTimes(0)
  expect(wrapper.vm.found).toBeFalsy()
})

test('Get Results - Error', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(ResultsTable, {
    mocks: {
      $route: { params: { event: 'EVENT' } },
      $messages: { addMessage: mockAddMessage },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockRejectedValue()
  await wrapper.vm.getResults()
  expect(wrapper.vm.rawResults).toEqual([])
  expect(mockAddMessage).toHaveBeenCalledTimes(0)
  expect(wrapper.vm.found).toBeFalsy()
})

test('Get Event Details - Makes Request with Correct ID', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(ResultsTable, {
    mocks: {
      $route: { params: { event: 'EVENT' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.resetAllMocks()
  axios.get.mockResolvedValue({ data: [] })
  await wrapper.vm.getEvent()
  expect(axios.get).toHaveBeenCalledTimes(1)
  expect(axios.get).toHaveBeenLastCalledWith('/api/events/EVENT')
})

test('Get Event Details - No Event Found', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(ResultsTable, {
    mocks: {
      $route: { params: { event: 'EVENT' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  expect(wrapper.vm.eventFound).toBeTruthy()
  axios.get.mockResolvedValue({ data: false })
  await wrapper.vm.getEvent()
  expect(wrapper.vm.eventFound).toBeFalsy()
})
