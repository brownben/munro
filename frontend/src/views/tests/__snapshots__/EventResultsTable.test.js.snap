// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly + Fetches Data 1`] = `
<div>
  <vue-headful-stub
    description="Results from the 200m event of the Justice league on Munro - League Results. Sorted. Sports League Results Calculated Quick and Easily, with Results Sorting and Filtering Options"
    head="[object Object]"
    title="Munro - 200m Event Results"
    url="https://munro-leagues.herokuapp.com/events/undefined"
  />
   
  <!---->
   
  <layout-stub
    title=""
  >
    <vue-headful-stub
      description="Page Not Found on Munro - League Results. Sorted. Sports League Results Calculated Quick and Easily, with Results Sorting and Filtering Options"
      head="[object Object]"
      title="Munro - I'm Lost"
    />
     
    <div
      class="w-full col-span-2 text-center"
    >
      <img
        alt="Person Lost looking at Map"
        class="inline-block w-full mt-8 md:w-2/3 xl:w-7/12 max-h-112"
        src="@/assets/images/UndrawLost.svg"
      />
       
      <h1
        class="mt-8 text-4xl font-bold text-main-900 md:mt-12 md:text-5xl leading-12"
      >
        
      Sorry, I'm Lost
    
      </h1>
       
      <h2
        class="mt-4 text-lg text-opacity-75 md:text-xl text-main-900"
      >
        
      I can't currently find what you are looking for, the link might be
      broken or the page may have moved.
    
      </h2>
       
      <p
        class="mt-8 text-lg text-opacity-75 md:text-xl md:mt-12 text-main-900 font-heading"
      >
        
      Try Looking Here:
    
      </p>
       
      <div
        class="mt-1"
      >
        <router-link-stub
          class="inline-block px-4 py-2 mx-2 mt-2 text-xl transition duration-300 bg-main-100 text-main-800 rounded-shape hover:bg-main-200"
          to="/"
        >
          Home
        </router-link-stub>
         
        <router-link-stub
          class="inline-block px-4 py-2 mx-2 mt-2 text-xl transition duration-300 bg-main-100 text-main-800 rounded-shape hover:bg-main-200"
          to="/leagues"
        >
          Leagues
        </router-link-stub>
         
        <router-link-stub
          class="inline-block px-4 py-2 mx-2 mt-2 text-xl transition duration-300 bg-main-100 text-main-800 rounded-shape hover:bg-main-200"
          to="/latest-results"
        >
          Latest Results
        </router-link-stub>
      </div>
    </div>
  </layout-stub>
</div>
`;
